### PR TITLE
fix(operations): G0.8-T5 document connector_id format + 404 on unknown (#630)

### LINKED ISSUE #630: Task: G0.8-T5 Document connector_id format and 404 on unknown — end the empty-200 "empty catalog" trap
Parent initiative: #634

> Originated as a consumer dogfood signal. The root-cause analysis below
> is consumer-supplied and was codebase-verified during the G0.8
> `/review-ticket` pass (branch `main` @ b5de86d, 2026-05-18).

## Target

`GET /api/v1/operations/groups` and `/api/v1/operations/search` require
a `connector_id` query param declared as a bare `str` with **no
description** — and an *unknown* `connector_id` returns an empty list +
**HTTP 200**, indistinguishable from "this connector has no operations".
A real dogfood evaluator concluded the catalog was empty when 40
descriptors were registered; the query was just shaped wrong
(`connector_id` is `<impl_id>-<version>`, not the product name) and
nothing said so. Document the format and fail loud on an unknown value.

## Root cause (consumer-supplied, codebase-verified)

- `backend/src/meho_backplane/api/v1/operations.py:66` and `:80` —
  both routes declare `connector_id: str = Query(min_length=1)` with no
  `description`/`examples`. The correct shape is `<impl_id>-<version>`
  (e.g. `vmware-rest-9.0`, `vault-1.x`), shown only in passing in
  `docs/cross-repo/connector-ingestion.md`.
- `backend/src/meho_backplane/operations/meta_tools.py:229-232` — a
  deliberate comment + code path returns an **empty `groups` list (HTTP
  200)** for an unknown `connector_id` ("empty is operationally
  meaningful"). That conflates "unknown connector" with "known
  connector, no enabled groups" — the trap.

## Files touched

- **Modify** `backend/src/meho_backplane/api/v1/operations.py:66,80` —
  add `description=` (+ `openapi_examples`) to both `connector_id`
  `Query(...)` declarations documenting the `<impl_id>-<version>`
  format with concrete examples.
- **Modify** `backend/src/meho_backplane/operations/meta_tools.py:229-232`
  (and the matching `/search` path) — distinguish *unknown connector*
  (raise `HTTPException(404)`) from *known connector, zero enabled
  groups* (keep `200 []`). Requires a cheap connector-exists check
  against the connector registry before returning the empty list.

## Approach

Lowest-friction correct fix, both asks (the consumer noted either alone
helps; both is ideal and the marginal cost of the second is one
registry lookup):

1. **Document the param** — `Query(..., description="Connector
   implementation id in `<impl_id>-<version>` form, e.g.
   `vmware-rest-9.0`, `vault-1.x` — NOT the bare product name",
   openapi_examples={...})`. Surfaces in the OpenAPI schema the agent
   and operators read.
2. **404 on unknown** — look the connector up in the existing connector
   registry; if absent, `raise HTTPException(status_code=404,
   detail="unknown connector_id '<x>' — expected <impl_id>-<version>;
   see GET /api/v1/connectors")`. A known connector with no enabled
   groups still returns `200 []` (that *is* meaningful). 404 chosen over
   422 because the value is well-formed but names a non-existent
   resource (REST-correct; 422 is for malformed input).

## Acceptance criteria

- [ ] The generated OpenAPI (`GET /openapi.json`) shows a non-empty
      `description` and at least one example for `connector_id` on both
      `/operations/groups` and `/operations/search`.
- [ ] `GET /api/v1/operations/groups?connector_id=vault` (bare product,
      unknown impl id) returns **404** with a `detail` naming the
      `<impl_id>-<version>` format — not `200 []`.
- [ ] A known connector with zero enabled groups still returns `200`
      with an empty list (regression guard — the "meaningful empty"
      case is preserved).
- [ ] `/operations/search` behaves identically for unknown vs
      known-empty.
- [ ] `pytest` covers unknown→404, known-empty→200[], valid→200[data];
      CI green; `ruff` + `mypy` clean.
- [ ] #634 Child-Tasks checklist line for #630 ticked.

## Out of scope

- Fuzzy-matching a bare product name to its impl id (e.g. `vault` →
  `vault-1.x`) — out of scope; the fix is to fail loud + document, not
  to guess the operator's intent.
- Reshaping the `/connectors` discovery surface — separate concern; the
  404 `detail` merely points at it.

## References

- Parent Initiative: [#634](https://github.com/evoila/meho/issues/634)
  (G0.8) — DoD line 5.
- Codebase sites:
  `backend/src/meho_backplane/api/v1/operations.py:66,80`,
  `backend/src/meho_backplane/operations/meta_tools.py:229-232`,
  `docs/cross-repo/connector-ingestion.md` (the `<impl_id>-<version>`
  precedent).
- Framework docs (pinned to repo, fastapi>=0.110): Query parameter
  metadata / `description` / `openapi_examples` —
  https://fastapi.tiangolo.com/tutorial/query-params-str-validations/ ;
  raising HTTP errors —
  https://fastapi.tiangolo.com/tutorial/handling-errors/ .
- Best-practices anchor: **Python best-practices pack** (fail-loud over
  ambiguous empty success; self-documenting API contracts) — cited by
  pack name (repo ships only `-slim` skills; anchor paths dead).

---

## MEHO triage

- **Parent initiative:** #634 (G0.8 v0.2 dogfood hardening).
- **Owner:** @damir-topic.
- **Priority tier:** High.
- **Labels:** `task, bug, connector, documentation, priority:high, effort:small`.
- **Best-practices anchor:** Python best-practices pack (by name).

### PR DESCRIPTION
**Pull request by damir-topic**
## Summary

- `GET /api/v1/operations/groups` and `/search` now document the
  `connector_id` format (`<impl_id>-<version>`, e.g. `vmware-rest-9.0`,
  `vault-1.x`) with a non-empty `description` and three named
  `openapi_examples` — the contract is now in the generated
  `/openapi.json` the agent and operators read, not just buried in
  `docs/cross-repo/connector-ingestion.md`.
- An **unknown** `connector_id` is now a **404** instead of an empty
  `200 []`. The empty-200 conflated "unknown connector" with "known
  connector, no enabled groups" — a real dogfood evaluator concluded
  the catalog was empty when 40 descriptors were registered; the id
  was just mis-shaped (bare product `vault` vs `vault-1.x`).
- A **known** connector with zero enabled groups still returns
  `200 []` (the "meaningful empty" case is preserved — explicit
  regression guard added).
- Design note: the unknown-connector signal is a domain exception
  (`UnknownConnectorError(ValueError)`) raised by the shared meta-tool
  handlers; the **REST route** maps it to 404. It is deliberately
  **not** `fastapi.HTTPException` raised from the handler — the same
  handlers back the MCP transport and CLI, and the MCP server's
  generic `except Exception` would mistranslate an `HTTPException`
  into a JSON-RPC `INTERNAL_ERROR` (a worse trap than the empty-200).
  This mirrors the existing `ValueError → HTTPException(400)` pattern
  already in `api/v1/operations.py`.

## Test plan

- [x] `ruff check` / `ruff format --check` — clean
- [x] `mypy src/meho_backplane/ --ignore-missing-imports` — clean
- [x] `code-quality.py --diff` — 0 blockers (3 pre-existing-debt
      warnings on `meta_tools.py` size, recorded as adjacent findings)
- [x] AC1: generated OpenAPI shows non-empty `description` (259 chars)
      + 3 examples for `connector_id` on both `/groups` and `/search`
      (verified by introspecting `app.openapi()`)
- [x] AC2: `connector_id=vault` (bare product) and `=ghost-9.9`
      (unknown) → 404 with `detail` naming the `<impl_id>-<version>`
      format (`test_get_groups_unknown_connector_returns_404`,
      `test_get_groups_bare_product_name_returns_404`)
- [x] AC3: known connector, zero enabled groups → `200 []`
      (`test_get_groups_known_connector_zero_enabled_returns_empty_200`)
- [x] AC4: `/search` behaves identically for unknown vs known-empty
      (`test_get_search_unknown_connector_returns_404`,
      `test_get_search_known_connector_no_match_returns_empty_200`)
- [x] AC5: full unit lane (`pytest -n auto --ignore=tests/integration`)
      green; stale acceptance assertions that encoded the old
      empty-200 contract updated to the new fail-loud contract

## Out of scope

- Fuzzy-matching a bare product name to its impl id (`vault` →
  `vault-1.x`) — the fix is fail-loud + document, not guessing intent.
- Reshaping the `/connectors` discovery surface — the 404 `detail`
  merely points at `GET /api/v1/connectors`.

Closes #630

## Summary by CodeRabbit

* **Bug Fixes**
  * Operations API and backend tools now consistently treat unknown connector identifiers as errors (HTTP 404) while returning HTTP 200 with empty results for known connectors that have no matching data.

* **Tests**
  * Expanded and updated tests to assert the new unknown-connector error semantics and tenant-scoping behavior across API and meta-tool layers.

* **Chores**
  * Updated ignore rules to exclude workspace files (*.code-workspace).

[Review Change Stack](https://app.coderabbit.ai/change-stack/evoila/meho/pull/649?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

---

## Follow-up (auto-implement-issue, iteration 1, 2026-05-18)

Addressed review findings from the iteration-1 `REQUEST_CHANGES` review:

- **B1** `a2ee899` — tenant-scope `connector_exists`: added a `tenant_id`
  parameter, applied `(tenant_id IS NULL) | (tenant_id == tenant_id)` to
  both `LIMIT 1` probes (mirroring `meta_tools.py:283-285`), passed
  `tenant_id=operator.tenant_id` from both callers. Added a multi-tenant
  regression test (`test_connector_existence_is_tenant_scoped`): a
  tenant-B-private connector reads as unknown (404) to tenant A, while a
  NULL-tenant shared connector stays visible to all. The pre-existing
  `test_search_operations_enforces_tenant_boundary` was updated to seed a
  shared descriptor so it still exercises the hit-list boundary rather
  than the now-correct per-tenant existence 404.

Disputed: none.

Deferred: none. (Pre-existing `meta_tools.py` size warnings remain out of
scope, as noted in the original PR body and the review.)

### CODE REVIEW

**Comment by damir-topic** [comment]
@coderabbitai review

File: backend/src/meho_backplane/operations/_lookup.py [hunk 1]
```diff
@@ -156,3 +157,52 @@ async def count_known_ops(
...
+        group_hit = await session.execute(
+            select(OperationGroup.id)
+            .where(
+                OperationGroup.product == product,
+                OperationGroup.version == version,
+                OperationGroup.impl_id == impl_id,
+            )
+            .limit(1)
+        )
+        return group_hit.first() is not None
```
Comment (damir-topic) [inline backend/src/meho_backplane/operations/_lookup.py:222]: **[🛑 B1]** — `connector_exists` ignores tenant scoping (also flagged by CodeRabbit)

**What:** `connector_exists` matches `endpoint_descriptor` / `operation_group` rows for `(product, version, impl_id)` with **no tenant filter**, while the data-returning queries it gates (`list_operation_groups`, `search_operations` in `meta_tools.py`) scope to `(tenant_id IS NULL) OR (tenant_id == operator.tenant_id)`.

**Why it matters:** Two concrete defects in MEHO's multi-tenant model:
1. Cross-tenant presence oracle — a connector private to tenant B (`tenant_id == B`) makes `connector_exists` return `True` for a tenant-A caller, so the route returns `200 []` instead of `404`. Tenant A learns the connector exists somewhere.
2. Contract regression — AC2/AC3 ("unknown → 404", "known-empty → 200 []") are computed against cross-tenant rows. For a caller who cannot see the connector, the caller-visible answer is "unknown" → must be `404`, but this yields `200 []` — the exact empty-catalog trap this task removes, just re-scoped per tenant. The existing single-tenant tests pass only because they never exercise a foreign-tenant-private connector.

**Required change:** add a `tenant_id: UUID` parameter to `connector_exists` and apply `(EndpointDescriptor.tenant_id.is_(None)) | (EndpointDescriptor.tenant_id == tenant_id)` (and the same on `OperationGroup`) to both `LIMIT 1` probes — mirroring the scoping already in `list_operation_groups`. Pass `tenant_id=operator.tenant_id` from both callers in `meta_tools.py`. Add a regression test: tenant A queries a tenant-B-private connector_id → `404` (and a tenant-A-visible-but-zero-enabled connector still → `200 []`).

**Source:** scoping shape per `meho_backplane/operations/meta_tools.py:283-285` (the `group_stmt` tenant filter this helper must match).

**Review by damir-topic — COMMENTED** [review]
## /auto-review-pr summary (iteration 1)




**Anchor task:** #630 — G0.8-T5 document connector_id format + 404 on unknown (end the empty-200 trap)
**Surface area:** backend
**Best-practices packs audited:** python
**Verdict:** REQUEST_CHANGES
**Head SHA reviewed:** `d0ea7b1`
**AI reviewer (CodeRabbit):** received in ~60s — 1 actionable (🟠 Major), confirmed valid and promoted to blocker

### Acceptance criteria
- [x] AC1 — OpenAPI documents `<impl_id>-<version>` on `/groups` and `/search`: verified by introspecting `app.openapi()` — 259-char `description` + 3 `openapi_examples` (`vmware-rest-9.0`, `vault-1.x`, `k8s-1.x`) on both routes.
- [❌] AC2 — unknown connector_id → 404: holds single-tenant; **not provably correct multi-tenant** (see B1) — a connector private to another tenant yields `200 []` for the caller, not `404`.
- [❌] AC3 — known-but-zero-enabled → 200 []: holds single-tenant; multi-tenant the unknown/known boundary is computed against cross-tenant rows (see B1).
- [x] AC4 — `/search` parity with `/groups`: identical `UnknownConnectorError` path + tests.
- [x] AC5 — pytest/ruff/mypy green: verified locally (34 passed in the two operations suites; 2 passed / 28 skipped acceptance; ruff + mypy clean).

### Predicted vs. actual diff
- ✅ `api/v1/operations.py` (Query metadata + `UnknownConnectorError`→404 mapping), `meta_tools.py` (domain exception raised by shared handlers), `mcp/tools/operations.py` (tool-description updates), test updates incl. stale-acceptance fixups — all expected and correct.
- ❓ `_lookup.py::connector_exists` — justified new helper, but **wrong** on tenant scoping (see B1, inline).
- ✅ `.gitignore` — `*.code-workspace`; trivial, disclosed.

### Design deviation (verified correct)
The worker did **not** follow the issue's literal "raise `HTTPException(404)` in `meta_tools.py`". It introduced `UnknownConnectorError(ValueError)` in the shared handlers and maps it to 404 only in the REST route. **Confirmed correct:** `mcp/server.py:495` has a generic `except Exception` rendering JSON-RPC `INTERNAL_ERROR`; a `fastapi.HTTPException` raised from the shared handler would be mistranslated for the MCP transport. The `ValueError → HTTPException(400)` precedent at `api/v1/operations.py:167-169` is real, so the new pattern mirrors existing code. Deviation accepted.

### Integration verification
- [x] `ruff check src/` — clean
- [x] `ruff format --check src/` — clean
- [x] `mypy` (operations + operations route) — clean
- [x] `pytest tests/test_api_v1_operations.py tests/test_operations_meta_tools.py` — 34 passed
- [x] `pytest` touched acceptance tests — 2 passed / 28 skipped
- [x] OpenAPI introspection via `meho_backplane.main:app` — AC1 confirmed
- ⚠️ not run — multi-tenant cross-visibility scenario: no test exercises tenant A querying a tenant-B-private connector (test gap mirrors the code gap in B1)

### Findings (inline — each is a merge blocker per team policy)
For the `/auto-implement-issue --continue-pr` punch list:

- 🛑 B1 — `connector_exists` ignores tenant scoping (also flagged by CodeRabbit); required change: add `tenant_id: UUID` param, filter both probes on `(tenant_id IS NULL) | (tenant_id == tenant_id)`, pass `operator.tenant_id` from both callers, add a multi-tenant regression test (tenant A → other-tenant-private connector → 404).

### Adjacent improvements (non-blocking, not posted inline)
- PR body notes 3 pre-existing `code-quality.py` debt warnings on `meta_tools.py` size — not introduced here; out of scope for this PR.

### AI reviewer (CodeRabbit) cross-check
- Adopted: 1 (B1 — tenant scoping; cited inline)
- Disputed: 0
- Out of scope: 0

### Docs
- [x] n/a — no `docs/codebase/` artifact for this slice; the contract is now self-documenting in generated OpenAPI (the task's explicit intent). PR body correctly notes the prior home (`docs/cross-repo/connector-ingestion.md`) is superseded by the OpenAPI surface, not invalidated.

### What worked well (review-body only)
- The deviation from the literal issue instruction is the *right* call and is documented precisely (PR body + `UnknownConnectorError` docstring), with the real MCP `except Exception` trap correctly diagnosed.
- Tests are honest: they assert the new fail-loud contract, the preserved meaningful-empty case, and `/search` parity — and the stale acceptance assertions that encoded the old empty-200 contract were updated rather than deleted.

**Comment by damir-topic** [comment]
@coderabbitai review

**Review by damir-topic — COMMENTED** [review]
## /auto-review-pr summary (iteration 2)




**Anchor task:** #630 — G0.8-T5 document connector_id format + 404 on unknown
**Surface area:** backend
**Best-practices packs audited:** python
**Verdict:** APPROVE
**Head SHA reviewed:** `a2ee899`
**AI reviewer (CodeRabbit):** not received within 15min (poll: 30 attempts / 875s). CodeRabbit's incremental model does not re-review already-reviewed commits, and the only `@coderabbitai review` trigger fired before the fix commit was pushed. Its iteration-1 Major on `_lookup.py:222` is the exact B1 and is directly resolved by this fix (which implements CodeRabbit's own proposed diff verbatim).

### Iteration-1 blocker B1 — resolution verified

`connector_exists` ignored tenant scoping (cross-tenant presence oracle + known-empty-vs-404 regression). All four B1 conditions verified resolved on `a2ee899`:

- [x] **`connector_exists` tenant-scoped** — `tenant_id: UUID` param added; both `EndpointDescriptor` and `OperationGroup` LIMIT-1 probes filter `(tenant_id IS NULL) | (tenant_id == tenant_id)`, mirroring `meta_tools.py:283-285`.
- [x] **Both callers pass `tenant_id=operator.tenant_id`** — `list_operation_groups` and `search_operations`.
- [x] **Multi-tenant regression test** — `test_connector_existence_is_tenant_scoped` proves (a) tenant-A-private connector → `UnknownConnectorError` (REST→404, not `200 []`) for tenant B, and (b) NULL-tenant shared connector visible to both tenants (no over-restriction false 404).
- [x] **Pre-existing test not gutted** — `test_search_operations_enforces_tenant_boundary` still asserts the hit-list boundary (tenant A sees `vault.tenant.a.op`, tenant B never does). The added shared `tenant_id IS NULL` descriptor only keeps the connector *known* to both tenants so the now-correct existence gate doesn't 404 tenant B before the hit-list assertion runs. Boundary remains meaningfully exercised.

### Acceptance criteria
- [x] AC1 — connector_id format documented + `openapi_examples` on `/groups` and `/search` (verified iteration 1; unchanged).
- [x] AC2 — unknown connector_id → 404 (not empty `200 []`); `UnknownConnectorError`→`HTTPException(404)` at `api/v1/operations.py:112-113, 145-146`. **Intact, not reverted** by the B1 fix.
- [x] AC3 — unknown-vs-known-empty decision is now provably correct multi-tenant (the iteration-1 gap that made AC3 non-provable).

### Project-board alignment
| Issue | Status | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #630 | Open | Closes | ✅ | ✅ | close on merge |
| #634 | Open | Parent initiative | — | n/a | no change |

### Predicted vs. actual diff
- ✅ `_lookup.py`, `meta_tools.py`, `test_operations_meta_tools.py` — all expected and correct; no undisclosed scope, nothing missing.

### Integration verification
- [x] `ruff check` / `ruff format --check` — clean (3 files already formatted)
- [x] `mypy` — clean, no issues in 33 source files
- [x] `pytest` — 41 passed (meta_tools + api_v1_operations + mcp_tools_operations); 3/3 tenant-boundary tests pass
- [⚠️ not run] migrations replay — no schema change in fix commit
- [x] multi-tenant visibility — now exercised by `test_connector_existence_is_tenant_scoped` (iteration-1 reasoned-only gap closed)

### Findings (inline)
None. Iteration-1 B1 fully resolved; no new findings.

### Adjacent improvements (non-blocking, not posted inline)
- Pre-existing `code-quality.py` debt (meta_tools.py file/function sizes; `connector_exists` 61-line docstring-dominated) — flagged out-of-scope in the original PR body and iteration-1 review; not introduced here.

### AI reviewer (CodeRabbit) cross-check
- Adopted: 1 (iteration-1 Major, promoted to B1, now resolved by this fix)
- Disputed: 0
- Out of scope: 0

### Docs
- [x] `docs/codebase/` — verified unchanged (fix is a visibility-scoping tightening; existing connector_exists docstring updated in-file to document the tenant boundary)

### What worked well
- The fix mirrors the existing `meta_tools.py:283-285` tenant idiom exactly rather than inventing a new pattern, and `connector_exists` already had a `count_known_ops` sibling taking `tenant_id: UUID`, so the signature change is consistent with the module contract.
- The regression test asserts both directions (private → 404 *and* shared → visible), and the pre-existing boundary test was preserved with a deliberate, documented seed change rather than weakened to pass.

---

> [!NOTE]
> Posted as a **comment** rather than a formal Approve: GitHub blocks `gh pr review --approve` on the author's own PR. The binary verdict is **APPROVE** and is authoritative from the result file. A human/second identity must click **Approve** to satisfy the branch ruleset's review requirement before merge.
